[Feat]: Migrate NFT Actions, Mutations and State

Jennievon opened the pull request.

Jennievon committed:
[Feat]: Migrate NFT Actions, Mutations and State

Jennievon committed:
fix: assign type for `NFTAssets` state

kraikov committed:
feat: integrate chainify

kraikov committed:
feat: rename CAL to chainify

kraikov committed:
feat: add approve step to all liquality swaps

kraikov committed:
refactor: set btc fee provider only for mainnet

kraikov committed:
chore: client & swap factory folder structure

kraikov committed:
fix: updateTransactionFee using the swap provider if possible

monokh commented on src/store/actions/getNFTAssets.js:
```diff
@@ -0,0 +1,38 @@
+
+  commit('SET_NFT_ASSETS_NUMBER', nft.assets.length)
+
+  let result = nft.assets.reduce(function (r, a) {
```
Name `r` and `a` so it's clearer what's happening

monokh commented on src/store/types.ts:
```diff
@@ -208,4 +208,126 @@ export interface RootState {
+  starredNFTs: NFTAsset[]
+}
+
+export type NFTAsset = {
```
Why is this type defined in wallet-core? Doesn't it come from CAL/chainify? It should be imported from there

monokh replied:
?

monokh commented on src/store/actions/getNFTAssets.js:
```diff
@@ -0,0 +1,38 @@
+    }
+  })
+
+  commit('SET_NFT_ASSETS_NUMBER', nft.assets.length)
```
There's no need to commit something to the state that can be derived from the state directly.
When you need the length you can do state.nfts.length

monokh commented on src/store/actions/getNFTAssets.js:
```diff
@@ -0,0 +1,38 @@
+  commit('SET_NFT_ASSETS_NUMBER', nft.assets.length)
+
+  let result = nft.assets.reduce(function (r, a) {
+    r[a.collection.name] = r[a.collection.name] || []
```
Sorting and grouping should not be in the action and is more of a UI concern.
Rather have all NFTs flat in the structure and sorting/ grouping happening elsewhere. You can add it as a getter if you want - there are similar ones in the getters you can look at

monokh commented on src/store/state.ts:
```diff
@@ -49,6 +49,9 @@ const initialState: RootState = {
   enabledChains: {},
+  nftAssets: {},
+  nftAssetsNumber: 0,
+  starredNFTs: []
```
There seems to be a lot of complexity introduced because of this array which is separately storing the starred NFTs. It would be better if `starred` was just a property of the NFT, to get a list of starred NFTs, you have to just filter that and that can be done on the UI

monokh commented on src/store/mutations.ts:
```diff
@@ -423,6 +425,34 @@ export default {
     };
   },
+  SET_NFT_ASSETS(state: RootState, payload: NFTAssets) {
+    state.nftAssets = payload
```
Aren't the nfts different depending on the network and walletid at the very least? what about depending on account? If so, then they need to be split in the state under those categories. See for example the `history` state object

monokh reviewed (changes requested):
Please also add a test for the actions that does some basic checks that nfts are retrieved

kraikov committed:
fix: swap providers & send transaction

Jennievon committed:
Merge branch 'master' of https://github.com/liquality/wallet-core into NFTs-migration

Jennievon committed:
[WIP]: NFT Send Transaction

Jennievon committed:
[WIP]: NFT Transfer method

Jennievon committed:
Merge branch 'master' of https://github.com/liquality/wallet-core into NFTs-migration

kraikov committed:
fix: send version string to atomic agent

kraikov committed:
chore: update dependencies to match the one from wallet

kraikov committed:
Merge branch 'master' into feat/chainify

Jennievon committed:
feat: add NFT to transaction types and history

Jennievon committed:
revert infuraApi URL

Jennievon committed:
feat: update methods for NFT Send txn details

kraikov committed:
chore: import from index

kraikov committed:
fix: liquality swaps

kraikov committed:
fix: approve should also work for liquality swaps

kraikov committed:
fix: use correct chain provider for btc

kraikov committed:
feat: add htlc address in the build config

kraikov committed:
fix: return timeline

kraikov committed:
Merge branch 'master' into feat/chainify

kraikov committed:
feat: chainify@1.0.3

Jennievon committed:
feat: add test for `getNFTAssets`

Jennievon committed:
refactor:
- move sorting and grouping to a getter function
- refactor getNFT method

monokh commented on src/factory/client.ts:
```diff
@@ -123,7 +124,7 @@ function createEthereumClient(
     ethClient.addProvider(
       new EthereumJsWalletProvider({
-        network: ethereumNetwork,
+        network: { ...ethereumNetwork, chainId: 4 },
```
Are we really switching to rinkeby? As mentioned this will break thorchain

monokh replied:
maybe you can share the rinkeby version with QA just for testing, but keep ropsten in the official build

Jennievon replied:
How do we handle it when we need to release though?

monokh replied:
I'm not sure we can support testnet NFTs with the current state of the wallet. The liquidity providers, liquality and thorchain are running on ropsten and we can't sacrifice that for NFTs

We have to have a separate conversation for how to enable multiple testnet networks for the wallet

AndonMitev replied:
I heard somewhere rinkeby will be deprecated

kraikov replied:
We should not switch to Rinkeby. Can you please revert this back.

monokh commented on src/factory/client.ts:
```diff
@@ -147,6 +148,9 @@ function createEthereumClient(
   // TODO: remove when Chainify is added
   (ethClient.wallet as any).signTypedMessage = signTypedMessage.bind(ethClient.wallet);
 
+  const openSeaAPI = 'https://rinkeby-api.opensea.io/api/v1/';
```
Only rinkeby? What about mainnet?

monokh commented on src/store/actions/performNextAction/sendNFT.ts:
```diff
@@ -0,0 +1,56 @@
+    const tx = await client.chain.getTransactionByHash(transaction.txHash);
+    console.log('🚀 ~ file: sendNFT.ts ~ line 30 ~ tx', tx);
+    if (tx && tx.confirmations && tx.confirmations > 0) {
+      dispatch.updateBalances({
```
Why do you need to update balances? Isn't that just for assets not NFTs?

monokh replied:
Should you update the NFTs of the account?

monokh commented on src/store/actions/sendNFTTransaction.ts:
```diff
@@ -0,0 +1,92 @@
+    network: Network;
+    accountId: AccountId;
+    walletId: WalletId;
+    contract: Address | string;
```
Shouldn't these values "contract", "Receiver" "tokenIDs" all come from the `nft` ? Having duplicates of the data can cause problems

Jennievon replied:
Yeah, just the `contract` and `tokenIDs` though but noted. 👌🏽

monokh commented on src/store/getters.ts:
```diff
@@ -282,4 +282,17 @@ export default {
+  nftAssetsByCollection(...context: GetterContext): NFTAsset[] {
+    const { state } = rootGetterContext(context);
+    const { nftAssets } = state;
+    const result = nftAssets.reduce(function (assets: any, asset: any) {
```
No `any` please, the type of these objects should be known

monokh commented on src/store/getters.ts:
```diff
@@ -282,4 +282,17 @@ export default {
+        return assetA.starred === assetB.starred ? 0 : assetA.starred ? -1 : 1;
+      });
+      return assets;
+    }, Object.create(null));
```
`Object.create(null)` vs `{}`?

monokh commented on src/store/types.ts:
```diff
@@ -135,6 +136,18 @@ export interface SendHistoryItem extends BaseHistoryItem {
+  nft: NFTAsset;
+  txHash: string;
+  accountId: AccountId;
+  fiatRate: number;
```
is this needed?

Jennievon replied:
By 'this', do you mean `fiatRate`?

monokh replied:
yes

monokh commented on src/store/types.ts:
```diff
@@ -135,6 +136,18 @@ export interface SendHistoryItem extends BaseHistoryItem {
+export interface NFTSendHistoryItem extends BaseHistoryItem {
+  type: TransactionType.NFT;
+  toAddress: string;
+  amount: string;
```
does NFT have an amount?

monokh commented on src/store/state.ts:
```diff
@@ -49,6 +49,8 @@ const initialState: RootState = {
   watsNewModalVersion: '',
   enabledChains: {},
+  nftAssets: [],
+  starredNFTs: [],
```
I left a comment about not requiring a separate state for starred NFTs and just using the property in `nftAssets`. You can again use a getter if you really need it in a list like this

monokh commented on src/store/mutations.ts:
```diff
@@ -427,6 +437,23 @@ export default {
   },
+  SET_NFT_ASSETS(state: RootState, payload: { nftAssets: NFTAsset[]; network: Network; walletId: WalletId }) {
+    // const walletHistory = state.history[payload.network]?.[payload.walletId];
+    state.nftAssets = payload.nftAssets;
```
Don't think this comment has been addressed yet: 
https://github.com/liquality/wallet-core/pull/6#discussion_r867830110

We have to figure out handling different chain/network/account now. Once the state is created, it will be difficult to migrate users to different state when we need support for it (which would be very soon)

Jennievon committed:
fix:
- remove `any` interface and replace with appropriate types
- remove duplicate pparams in send NFT txn
- remove amount from NFT txn

AndonMitev commented on src/store/actions/getNFTAssets.ts:
```diff
@@ -0,0 +1,42 @@
+  const client = getters.client({
+    network,
+    walletId,
+    asset: 'ETH',
```
What if we want to get for polygon for example

AndonMitev commented on src/store/actions/getNFTAssets.ts:
```diff
@@ -0,0 +1,42 @@
+      nftAssets.push(asset);
+    } else {
+      asset['starred'] = false;
+      nftAssets.push(asset);
```
you can move `nftAssets.push` outside the if else because it's executed anytime anyway

AndonMitev commented on src/store/actions/getNFTAssets.ts:
```diff
@@ -0,0 +1,42 @@
+      nftAssets.push(asset);
+    }
+  });
+  console.log('🚀 ~ file: getNFTAssets.ts ~ line 34 ~ nftAssets ~ nftAssets', nftAssets);
```
`console.log`

AndonMitev commented on src/store/actions/performNextAction/sendNFT.ts:
```diff
@@ -0,0 +1,50 @@
+  });
+  try {
+    const tx = await client.chain.getTransactionByHash(transaction.txHash);
+    console.log('🚀 ~ file: sendNFT.ts ~ line 30 ~ tx', tx);
```
log

Jennievon committed:
refactor:
- remove unnecessary params

kraikov committed:
feat: chainify@1.0.4

kraikov committed:
chore: syntax sugar

kraikov committed:
Merge branch 'master' into feat/chainify

bradleySuira reviewed (approved):
LGTM

Jennievon committed:
refactor:
- rewrite logic for starring NFTs

monokh commented on src/store/actions/getNFTAssets.ts:
```diff
@@ -22,20 +22,50 @@ export const getNFTAssets = async (
-        ? true
-        : false;
-      nftAssets.push(asset);
+  const oldAssets: NFTAsset[] = state.nftAssets;
```
What is the purpose of this function?  Can you describe input and outputs?
It's really hard to understand and I feel there is a much clearer way to write what this is doing

Jennievon committed:
set nftAssets by

Jennievon committed:
feat:
- sort NFT assets by collection, walletId and accountId
- add NFTAssets to accounts state

Jennievon committed:
refactor:
- change variable names for  and add comments

Jennievon committed:
fix:
- update comment

kraikov committed:
Merge branch 'master' into feat/chainify

Jennievon committed:
feat:
- getter methods for  by collection (all accounts) & account

monokh commented on src/store/actions/getNFTAssets.ts:
```diff
@@ -0,0 +1,82 @@
+  const nftAssetsStoredInState: NFTAsset[] = account?.nftAssets || [];
+  const nftAssetsFetched: NFTAsset[] = nft.assets;
+
+  // set new nft assets to a new map
```
I see that you added comments but we really need to simplify this function overall
It looks like it's jsut merging 2 arrays and default starred = false. That shouldn't require this much code and should be doable with some map+reduce. Can you help @kraikov ?

kraikov replied:
I think we can simply use the fetched array and set the starred ones from the state.

Jennievon replied:
Thanks, @monokh @kraikov. 👍 
This has been addressed.

Jennievon committed:
refactor: rewrite logic for  action

kraikov commented on src/store/actions/getNFTAssets.ts:
```diff
@@ -0,0 +1,45 @@
+
+  const nftAssets: NFTAsset[] = [];
+
+  nftAssetsFetched.forEach((nftAsset) => {
```
can you use `map` instead of `forEach`?

kraikov commented on src/store/actions/sendNFTTransaction.ts:
```diff
@@ -0,0 +1,80 @@
+    nft: NFTAsset;
+  }
+): Promise<any> => {
+  const asset = 'ETH';
```
this one is confusing.... sending an NFT should not contain any `from` and `to` assets. Can we set them to `null` ?

kraikov replied:
can we pass the asset as function param, because in the future we would have different NFTs (e.g. Solana)

kraikov commented on src/store/actions/sendNFTTransaction.ts:
```diff
@@ -0,0 +1,80 @@
+    feeLabel,
+    nft,
+  }: {
+    network: Network;
```
Can we please have a dedicated type for the input param?

kraikov commented on src/store/actions/sendNFTTransaction.ts:
```diff
@@ -0,0 +1,80 @@
+    feeLabel: FeeLabel;
+    nft: NFTAsset;
+  }
+): Promise<any> => {
```
Avoid using `any`

kraikov commented on src/store/getters.ts:
```diff
@@ -282,4 +292,36 @@ export default {
+    // spread all nfts into one array
+    Object.values(nftAssetsByAccount).forEach((nftAssets) => {
+      nftAssetsByCollection = {
+        ...nftAssetsByCollection,
```
`nftAssetsByCollection` is always an empty array, right?

Jennievon replied:
For now but when we have NFTs in different accounts, we would still need to spread them into the array.

kraikov replied:
oh.. got it... then it will be better to use `reduce` and use different name for the function and parameter because it's a little bit confusing

kraikov commented on src/store/getters.ts:
```diff
@@ -282,4 +292,36 @@ export default {
+              return assetA.starred === assetB.starred ? 0 : assetA.starred ? -1 : 1;
+            });
+            return assets;
+          }, Object.create({}));
```
can't you just use `{}`

kraikov replied:
This is not yet resolved

kraikov commented on src/store/getters.ts:
```diff
@@ -282,4 +292,36 @@ export default {
+      if(account.nftAssets) {
+       const result = account.nftAssets.reduce(function (assets: NFTAssets, asset: NFTAsset) {
+            assets[asset.collection.name] = assets[asset.collection.name] || [];
+            assets[asset.collection.name].push(asset);
```
the above 2 lines can be combiled with:

`(assets[asset.collection.name] ||= []).push(asset)`

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_OR_assignment

kraikov reviewed (changes requested):
`package-lock.json` must not be included, because wallet-core uses `yarn.lock`

kraikov commented on src/store/actions/sendNFTTransaction.ts:
```diff
@@ -0,0 +1,80 @@
+    accountId,
+    feeLabel,
+  };
+  console.log('🚀 ~ file: sendNFTTransaction.ts ~ line 64 ~ transaction', transaction);
```
Remove.

Jennievon committed:
fix: address comments

Jennievon committed:
feat: add type check for send txn

Jennievon committed:
feat: use  function to set

Jennievon committed:
remove

Jennievon committed:
Merge branch 'master' of https://github.com/liquality/wallet-core into NFTs-migration

Jennievon committed:
fix:
- fix failing build test

kraikov commented on src/store/actions/getNFTAssets.ts:
```diff
@@ -0,0 +1,45 @@
+    if (nftAssetStoredInState) {
+      nftAsset.starred = nftAssetStoredInState.starred;
+    }
+    nftAssets.push(nftAsset);
```
the map function populates a new array with the returned value, so there is no need to do 
`nftAssets.push(nftAsset);`

You should do:

`const nftAssets =  nftAssetsFetched.map((nftAsset) => .... return nftAsset)`

kraikov commented on src/store/getters.ts:
```diff
@@ -298,4 +308,34 @@ export default {
+    const { getters } = rootGetterContext(context);
+    const { nftAssetsByAccount } = getters;
+    let nftAssetsByCollection: NFTAsset[] = [];
+    nftAssetsByCollection = Object.values(nftAssetsByAccount).reduce((accum, nftAssets) => {
```
The `reduce` function is not used correctly here. Please check https://codeburst.io/how-to-reduce-array-to-a-single-type-11cfe2bcae66

kraikov replied:
The `reduce` function is not used correctly here. Please check https://codeburst.io/how-to-reduce-array-to-a-single-type-11cfe2bcae66

Jennievon committed:
feat:
- refactor getters for nft assets
- pass asset in send txn from

Jennievon committed:
Update getters.ts

Jennievon committed:
Update types.ts

bradleySuira committed:
Feat/chainify ledger (#54)

* fix: send fee format based on native asset (#48)

* fix: send fee format based on native asset

* fix: getTxFee error

* hotfix: enable all default accounts including RSK legacy

* fix: trigger semantic release

* chore: updated nvmrc

* feat: updated implementation for ledger chainify

* fix: removed local resolutions

* feat: chainify@1.0.6

Co-authored-by: Slav Keremidchiev <skeremidch@gmail.com>
Co-authored-by: Krasimir Raykov <raykov@weichain.tech>

kraikov committed:
feat: chainify@1.0.7

kraikov committed:
fix: polygon & terra testnet urls

kraikov committed:
fix: disable mulicall for testnets

kraikov committed:
feat: chainify@1.0.8

bradleySuira committed:
updated packages

Jennievon committed:
feat:
fetch NFTs per network, per chain using the native asset

Jennievon committed:
feat: update types for NFTAssets

Jennievon committed:
Merge branch 'NFTs-migration' of https://github.com/liquality/wallet-core into NFTs-migration

kraikov committed:
feat: chainify@1.0.10

kraikov committed:
chore: replace console.error with console.info

skeremidchiev committed:
fix: lower case address on allowance (#56)

Jennievon committed:
Merge branch 'feat/chainify' of https://github.com/liquality/wallet-core into NFTs-migration

kraikov committed:
fix: do not pass network to formatAddress to satisfy RSK checksum

kraikov committed:
feat: chainify@1.0.11

Jennievon committed:
feat: update changes from chainify

Jennievon committed:
feat:
- update  type
- add methods to get NFT transfer and explorer link

Jennievon committed:
Merge branch 'master' of https://github.com/liquality/wallet-core into NFTs-migration

monokh commented on src/factory/client/clients.ts:
```diff
@@ -87,11 +87,14 @@ export function createEVMClient(
   // disable multicall for Testnets
   const chainProvider = new EvmChainProvider(ethereumNetwork, undefined, feeProvider, !ethereumNetwork.isTestnet);
   const swapProvider = new EvmSwapProvider({ contractAddress: HTLC_CONTRACT_ADDRESS });
+  const walletOptions = { derivationPath, mnemonic };
```
Unrelated change?

monokh commented on src/factory/client/evm.ts:
```diff
@@ -1,25 +1,33 @@
   const feeProvider = new EIP1559FeeProvider(ethNetwork.rpcUrl as string);
-  return createEVMClient(ethNetwork, feeProvider, mnemonic, accountType, derivationPath);
+  const chainProvider = new EvmChainProvider(
+    { ...ethNetwork, chainId: 4 },
```
As mentioned before, please let the wallet stay on ropsten testnet

Jennievon replied:
This wouldn't be merged. It's just to test the NFTs on testnet.

skeremidchiev committed:
Fix/add approve to LSP & Boosts (#57)

* fix: use sovryn amm as default

* fix: LSP and EVM SP steps

* fix: boost erc20 to native

* fix: boost native to erc20

monokh commented on src/store/actions/getNFTAssets.ts:
```diff
@@ -0,0 +1,47 @@
+  const { state, commit, getters } = rootActionContext(context);
+  const { networks, nftAssets } = buildConfig;
+
+  networks.forEach((network) => {
```
We shouldn't update for all networks when the user is not on it. It will slow down the wallet for no reason, 99% only use mainnet. the active network should be passed into this function

monokh commented on src/store/actions/getNFTAssets.ts:
```diff
@@ -0,0 +1,47 @@
+  const { networks, nftAssets } = buildConfig;
+
+  networks.forEach((network) => {
+    const assetKeys = nftAssets[network];
```
assetKeys? Isn't it chains?

monokh commented on src/store/actions/getNFTAssets.ts:
```diff
@@ -0,0 +1,47 @@
+    const assetKeys = nftAssets[network];
+    assetKeys.forEach(async (chain) => {
+      const asset = chains[chain].nativeAsset;
+      const account = state.accounts[walletId]![network].find((a) => a.assets.includes(asset));
```
Why do you need to find the account by the asset? You are looking for the account for that chain and accounts have a `chain` property

monokh replied:
There can be multiple accounts for a chain, each account (therefore address) has it's own list of nfts. You need to make sure every account is processed

monokh commented on src/store/actions/getNFTAssets.ts:
```diff
@@ -0,0 +1,47 @@
+        accountId: account?.id,
+      });
+
+      const nft: any = await _client.nft?.fetch();
```
why do we type this as any? I believe it does return a type

Jennievon replied:
`void` is being set as the type in `chainify` for the `fetch` response, hence, we cannot set a specific type as it isn't assignable to `void`. That's the reason I used `any`

monokh replied:
Yea it's not good that chainify is returning any 
Is the actual type `NftAsset` that you have in this project? In that case we can move it to chanify so you get the type here

monokh commented on src/store/actions/getNFTAssets.ts:
```diff
@@ -0,0 +1,47 @@
+        const nftAssetStoredInState = nftAssetsStoredInState.find((asset) => asset.id === nftAsset.id);
+        if (nftAssetStoredInState) {
+          nftAsset.starred = nftAssetStoredInState.starred || false;
+          nftAsset.asset = nftAssetStoredInState.asset || asset;
```
These 2 statements seem weird, asset and chain are static aren't they? they should always be the same so why the  || ?
Also, why store asset? an nft is for a chain, doesn't have anything to do with an asset i believe

monokh commented on src/store/migrations/before_after_migration_script.test.ts:
```diff
@@ -0,0 +1,116 @@
+import * as Process from 'process';
```
unrelated?

Jennievon replied:
some file changes are from the `master` branch

monokh replied:
does that mean chainify branch needs to be updated to include something? This PR should not be adding anything unrelated

Jennievon committed:
feat:
- only fetch NFTs by activeNetwork

Jennievon committed:
Merge branch 'feat/chainify' into NFTs-migration

Jennievon committed:
fix: merge conflicts

Jennievon committed:
feat:
sort account by chain

Jennievon committed:
Merge branch 'master' of https://github.com/liquality/wallet-core into NFTs-migration

Jennievon committed:
fix: remove unrelated changes

monokh committed:
fix: update chainify to get nft types - update star to toggle, update assets per account

monokh committed:
fix: return nft collections by account

Jennievon committed:
fix:
- update token_id to match number type
- update typo

monokh committed:
fix: handle multiple nft providers

monokh committed:
Merge branch 'NFTs-migration' of github.com:liquality/wallet-core into NFTs-migration

monokh committed:
fix: nft transfer

monokh committed:
fix: pass correct nft providers

monokh committed:
fix: arbitrum nft support + update moralis key

monokh committed:
fix: remove asset from nft send

monokh committed:
fix: return all nfts with account id

Jennievon committed:
feat:
- add transfer link for polygon
- filter NFT by token_id

Jennievon committed:
fix:
remove the unary plus operator on

monokh committed:
feat: update moralis providers

monokh committed:
fix: remove unneeded package

monokh commented:
Closing to reopen because github actions behaving funny
